src/gsi/stpcalc.f90

```diff
@@ -844,10 +844,10 @@ subroutine stpcalc(stpinout,sval,sbias,dirx,dval,dbias, &
            if(outpen(i) < outpensave)then
               stp(ii)=outstp(i)
               outpensave=outpen(i)
-              istp_use=i
+              istp_use=ii
            end if
         end do
-        if(istp_use /= istp_iter) then
+        if(istp_use /= nsteptot) then
```

Comment:
@XuLu-NOAA  Thanks for identifying the issue in , I believe, rarely occurring situations, I am also waiting for John 's thorough reviewing of this issue.  For being now, would you give some details what the issue is you identified when GSI debug mode is used.    From my current understanding of this part,  your change on line 850 looks "probably" right (when the step size at one previous step is selected, exit this loop), but the change on line 847 is suspicious. Just consider the "coding" itself (don't consider what the algorithm should be), only the change on line 850 could not resolve the problem encountered using debug mode GSI?

Reply:
Yes, @TingLei-NOAA . 
The error happens in an extreme situation where the optimal stepsize is not found after ii loop from 1 to istp_iter (which is set to 5). 

Then it will go into this part of code to look for the historical minimum outpen, and use the corresponding stp to be the final output for this search. 

In the original code, ii=istp_iter (which is 5) according to the if statement at L839. 
Yet L847 assigns istp_use=i (7 in this case; i loops from 1 to nsteptot, where nsteptot = ii*3+1 according to L704-710), while storing the minimum outstp(i) to stp(ii) at L845.
Then at L872, it tries to assign stp(istp_use) to stpinout, an error happened as stp is defined to be the size of istp_iter at L276, but istp_use is 7 in this case.

I believe the right logic is that since the minimum outstp is assigned to stp(ii), then istp_use should be ii as well. So the output stpinout can be using the right value at L872, even when i<istp_iter.

Reply:
@XuLu-NOAA 

After looking at the code carefully, I agree with your 2 changes.  Thanks for finding an issue that I probably introduced.